torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp

```diff
@@ -1236,6 +1239,15 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // hand over these tensors to WorkNCCL's stash when coalescing is ended.
   std::vector<at::Tensor> coalescedTensors_;
 
+  // Some ops may have completed, but user still hasn't called `work.wait()`.
+  // When watchdog detects this, it transfers the TensorShelf from `work` to
+  // this `shelves` structure. Next time we execute ProcessGroupNCCL's methods
+  // on main thread, we clear the `shelves` in one shot. This is mainly because
+  // watchdog (a side thread) unstashing the shelf directly seems to cause some
+  // problem.
+  std::vector<std::shared_ptr<TensorShelf>> shelvesToUnstash_;
+  std::mutex shelvesMutex_;
```

Comment:
nit: can we put the mutex before the variable it's protecting?

Reply:
Thanks. Addressed in https://github.com/pytorch/pytorch/pull/150130